posts/2020/crud-pattern-in-react-implementation.md

```diff
@@ -0,0 +1,130 @@
+.. title: CRUD Pattern in React - Implementation
+.. slug: crud-pattern-in-react-implementation
+.. date: 2020-07-08 12:33:18 UTC+05:30
+.. tags: ReactJS
+.. category: Mayur Borse
+.. link:
+.. description:
+.. type: text
+.. status: draft
+.. summary: This is a continuation post of our <strong>CRUD Pattern</strong> series. We introduced the pattern with the <em>Books</em> (parent) component in the <a href="/blog/crud-pattern-in-react-introduction">previous post</a>. This post provides the implementation details of <em>BookForm</em> and <em>BookTable</em> (child) components.
+
+# Overview
+
+In the <a href="/blog/crud-pattern-in-react-introduction">previous post</a>, we checked the structure of `Books` component that implements CRUD actions and passes these actions and data (retrieved from API) to its child components `BookTable` and `BookForm`. In this post, We will check implementation details of `BookTable` and `BookForm` components.
+
+# BookTable Component - Implementation
+
+A table view is generally used to provide summary of the data. It may also provide an interface (eg. button) to trigger _Update_ and _Delete_ actions on the selected data record. The `BookTable` component presents book entities in a summary view in table format. It triggers _Update_ and _Delete_ actions using `showEditForm` and `performDelete` callback functions, passed as `props` by the `Books` component (parent). It also passes the data required for the actions (book object, book id) back to the parent component.
+
+React supports Unidirectional/One-Way Data Flow only. It means, the data is passed from parent to child as `props` which are immutable arguments. Our actions (update, delete) are defined in the `Books` component (parent), but we need the data (selected book id or object) from `BookTable` component (child) to perform the actions. We can achieve this by explicitly mapping Inverse Data Flow from `BookTable` component to `Books` component.
+
+## Inverse Data Flow
+
+- A callback function (`openForm()` or `delete()`) defined in the parent (`Books`) component is passed as props to the child (`BookTable`) component.
+- The function is invoked by the child component for an event .
+- eg. to trigger update or delete action (function invoke) on button click (event).
+- The child component can pass data (book id or book object) as an argument to the callback function. As the callback function is defined in the parent, the data is made available to the parent for further action.
+- As data is passed in the _inverse_ direction (from child to parent), Inverse Data Flow is achieved.
+- The passed data can then be used by the parent component to update internal state which may potentially re-render the page.
+
```

Comment:
Let's add a small paragraph here that says. 

`BookTable` component is fairly straight forward. Here, we have covered important concepts in the `BookTable` component, for complete implementation [refer to](add link to the file in the repository). Next we will look at the `BookForm` component.